### PR TITLE
Cleanup "slow-path" logic in scheduler Filters

### LINKED ISSUE #87783: Cleanup "slow-path" logic in scheduler Filters
by "slow-paths"  we mean logic that allows executing Filters in the absence of PreFilter state. 

In the scheduler, it is not possible that a Filter will be invoked without invoking its PreFilter first, but we still have logic in Filters to allow this because historically ClusterAutoscaler didn't consistently call "MetadataProducer", neither kubelet and daemonset controller.

Now that both Kubelet and DaemonSet controller dependencies have been removed, and according to  https://github.com/kubernetes/kubernetes/pull/87702#issuecomment-581298101, CA is consistently calling PreFilter, this special logic can be removed.

[ ] https://github.com/kubernetes/kubernetes/blob/845b23232125caeb67eb30659ff482404a3d3735/pkg/scheduler/framework/plugins/interpodaffinity/filtering.go#L574

[ ] https://github.com/kubernetes/kubernetes/blob/845b23232125caeb67eb30659ff482404a3d3735/pkg/scheduler/framework/plugins/interpodaffinity/filtering.go#L411

[ ] https://github.com/kubernetes/kubernetes/blob/845b23232125caeb67eb30659ff482404a3d3735/pkg/scheduler/framework/plugins/nodeports/node_ports.go#L111

[ ] https://github.com/kubernetes/kubernetes/blob/845b23232125caeb67eb30659ff482404a3d3735/pkg/scheduler/framework/plugins/noderesources/fit.go#L158

/sig scheduling
/help
/cc @losipiuk @MaciekPytel @Huang-Wei

### PR DESCRIPTION
**Pull request by notpad**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Based on the discussion in #87702 and #87783, Prefilter will always be invoked before Filter, so state should be computed and wrote to CycleState by Prefilter, it won't be nil now.
 
This PR does two things:
- return an error if expected states don't exist in CycleState
- remove the "slow-path" logic which was used to handle nil state case

**Which issue(s) this PR fixes**:

Fixes #87783

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/sig scheduling
/assign @ahg-g

### CODE REVIEW

**Comment by notpad** [comment]
code is ready for review

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering.go [hunk 1]
```diff
@@ -388,10 +388,9 @@ func (pl *InterPodAffinity) RemovePod(ctx context.Context, cycleState *framework
 func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, error) {
 	c, err := cycleState.Read(preFilterStateKey)
 	if err != nil {
-		// The preFilterState wasn't pre-computed in prefilter. We ignore the error for now since
-		// Filter is able to handle that by computing it again.
+		// The preFilterState wasn't pre-computed in prefilter.
 		klog.V(5).Infof("Error reading %q from cycleState: %v", preFilterStateKey, err)
-		return nil, nil
+		return nil, err
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/interpodaffinity/filtering.go]:
```suggestion
		return nil, fmt.Errorf("error reading %q from cycleState: %v", preFilterStateKey, err)
```

And remove the above line.

File: pkg/scheduler/framework/plugins/nodeports/node_ports.go [hunk 2]
```diff
@@ -86,9 +87,9 @@ func (pl *NodePorts) PreFilterExtensions() framework.PreFilterExtensions {
 func getPreFilterState(cycleState *framework.CycleState) (preFilterState, error) {
 	c, err := cycleState.Read(preFilterStateKey)
 	if err != nil {
-		// preFilterState state doesn't exist. We ignore the error for now since
-		// Filter is able to handle that by computing it again.
-		return nil, nil
+		// preFilterState state doesn't exist.
+		klog.V(5).Infof("Error reading %q from cycleState: %v", preFilterStateKey, err)
+		return nil, err
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/nodeports/node_ports.go]: Ditto. We can simply return a verbose err and remove the logging.

File: pkg/scheduler/framework/plugins/noderesources/fit.go [hunk 3]
```diff
@@ -130,10 +130,9 @@ func (f *Fit) PreFilterExtensions() framework.PreFilterExtensions {
 func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, error) {
 	c, err := cycleState.Read(preFilterStateKey)
 	if err != nil {
-		// The preFilterState doesn't exist. We ignore the error for now since
-		// Filter is able to handle that by computing it again.
+		// The preFilterState doesn't exist.
 		klog.V(5).Infof("Error reading %q from cycleState: %v", preFilterStateKey, err)
-		return nil, nil
+		return nil, err
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/noderesources/fit.go]: Ditto. We can simply return a verbose err and remove the logging.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 4]
```diff
@@ -194,10 +194,9 @@ func (pl *PodTopologySpread) RemovePod(ctx context.Context, cycleState *framewor
 func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, error) {
 	c, err := cycleState.Read(preFilterStateKey)
 	if err != nil {
-		// The preFilterState wasn't pre-computed in prefilter. We ignore the error for now since
-		// we are able to handle that by computing it again (e.g. in Filter()).
+		// The preFilterState wasn't pre-computed in prefilter.
 		klog.V(5).Infof("Error reading %q from cycleState: %v", preFilterStateKey, err)
-		return nil, nil
+		return nil, err
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]: Ditto. We can simply return a verbose err and remove the logging.

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go [hunk 5]
```diff
@@ -192,10 +191,9 @@ func (pl *ServiceAffinity) RemovePod(ctx context.Context, cycleState *framework.
 func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, error) {
 	c, err := cycleState.Read(preFilterStateKey)
 	if err != nil {
-		// The metadata wasn't pre-computed in prefilter. We ignore the error for now since
-		// Filter is able to handle that by computing it again.
-		klog.V(5).Infof(fmt.Sprintf("reading %q from cycleState: %v", preFilterStateKey, err))
-		return nil, nil
+		// The metadata wasn't pre-computed in prefilter.
+		klog.V(5).Infof(fmt.Sprintf("Error reading %q from cycleState: %v", preFilterStateKey, err))
+		return nil, err
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]: Ditto. We can simply return a verbose err and remove the logging.

File: pkg/scheduler/framework/v1alpha1/cycle_state.go [hunk 6]
```diff
@@ -23,7 +23,7 @@ import (
 
 const (
 	// NotFound is the not found error message.
-	NotFound = "not found"
+	NotFound = "target key not found in CycleState"
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/v1alpha1/cycle_state.go]: Suggest to revert it back.

**Review by Huang-Wei — COMMENTED** [review]
Some nits, otherwise LGTM.

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering.go (hunk 1 above)
Comment (notpad) [reply pkg/scheduler/framework/plugins/interpodaffinity/filtering.go]: Thanks @Huang-Wei, done.

**Comment by Huang-Wei** [comment]
/approve

Will leave /lgtm to @ahg-g.

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go [hunk 7]
```diff
@@ -192,10 +190,8 @@ func (pl *ServiceAffinity) RemovePod(ctx context.Context, cycleState *framework.
 func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, error) {
 	c, err := cycleState.Read(preFilterStateKey)
 	if err != nil {
-		// The metadata wasn't pre-computed in prefilter. We ignore the error for now since
-		// Filter is able to handle that by computing it again.
-		klog.V(5).Infof(fmt.Sprintf("reading %q from cycleState: %v", preFilterStateKey, err))
-		return nil, nil
+		// The metadata wasn't pre-computed in prefilter.
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go]:
```suggestion
		// preFilterState doesn't exist, likely PreFilter wasn't invoked.
```

File: pkg/scheduler/framework/plugins/noderesources/fit_test.go [hunk 8]
```diff
@@ -398,64 +398,6 @@ func TestEnoughRequests(t *testing.T) {
 	}
 }
 
-func TestPreFilterDisabled(t *testing.T) {
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/noderesources/fit_test.go]: We should keep this test and verify that an error is returned when invoking Filter without calling PreFilter. The unit test should be simple.

File: pkg/scheduler/framework/plugins/nodeports/node_ports_test.go [hunk 9]
```diff
@@ -163,45 +163,6 @@ func TestNodePorts(t *testing.T) {
 	}
 }
 
-func TestPreFilterDisabled(t *testing.T) {
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/nodeports/node_ports_test.go]: keep the test and verify that Filter returns error.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 10]
```diff
@@ -194,10 +194,8 @@ func (pl *PodTopologySpread) RemovePod(ctx context.Context, cycleState *framewor
 func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, error) {
 	c, err := cycleState.Read(preFilterStateKey)
 	if err != nil {
-		// The preFilterState wasn't pre-computed in prefilter. We ignore the error for now since
-		// we are able to handle that by computing it again (e.g. in Filter()).
-		klog.V(5).Infof("Error reading %q from cycleState: %v", preFilterStateKey, err)
-		return nil, nil
+		// The preFilterState wasn't pre-computed in prefilter.
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go]:
```suggestion
		// preFilterState doesn't exist, likely PreFilter wasn't invoked.
```

File: pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go [hunk 11]
```diff
@@ -1634,124 +1634,6 @@ func TestRequiredAffinityMultipleNodes(t *testing.T) {
 	}
 }
 
-func TestPreFilterDisabled(t *testing.T) {
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/interpodaffinity/filtering_test.go]: Keep the test and verify that Filter now returns an error (it should be a single test case).

File: pkg/scheduler/core/generic_scheduler_test.go [hunk 12]
```diff
@@ -1339,7 +1339,7 @@ func TestSelectNodesForPreemption(t *testing.T) {
 			name: "a pod that fits on both machines when lower priority pods are preempted",
 			registerPlugins: []st.RegisterPluginFunc{
 				st.RegisterQueueSortPlugin(queuesort.Name, queuesort.New),
-				st.RegisterFilterPlugin(noderesources.FitName, noderesources.NewFit),
+				st.RegisterPluginAsExtensions(noderesources.FitName, 1, noderesources.NewFit, "Filter", "PreFilter"),
```
Comment (ahg-g) [inline pkg/scheduler/core/generic_scheduler_test.go]: The weight here is confusing, in a followup PR I suggest we change ```RegisterPluginAsExtensions``` to not set the weight, and have another function ```RegisterPluginAsExtensionsWithWeight``` that accept a weight parameter for the extensions that require a weight.

File: pkg/scheduler/framework/plugins/noderesources/fit_test.go (hunk 8 above)
Comment (notpad) [reply pkg/scheduler/framework/plugins/noderesources/fit_test.go]: Done

File: pkg/scheduler/core/generic_scheduler_test.go (hunk 12 above)
Comment (notpad) [reply pkg/scheduler/core/generic_scheduler_test.go]: I'll send another PR for this.

**Pull request merged**